File: prototypes/docs-rag-qa/typescript/src/api/query.ts
/**
 * Query route handler.
 */

import { Hono } from "hono";
import { answerQuestion } from "../agent/qa.js";
import { QueryRequest } from "../schemas/index.js";

export const queryRouter = new Hono();

queryRouter.post("/query", async (c) => {
  const body = await c.req.json();
  const parsed = QueryRequest.safeParse(body);

  if (!parsed.success) {
    return c.json(
      { error: "Invalid request", details: parsed.error.issues },
      400,
    );
  }

  const { question, top_k } = parsed.data;
  const traceId = crypto.randomUUID();

  const { text } = await answerQuestion(question, top_k);

  return c.json({
    answer: text,
    citations: [],
    trace_id: traceId,
  });
});

File: prototypes/docs-rag-qa/python/app/api/query.py
"""Query route handlers."""

import uuid

import structlog
from fastapi import APIRouter

from app.agent.qa import answer_question
from app.models.schemas import QueryRequest, QueryResponse

logger = structlog.get_logger()

router = APIRouter()


@router.post("/query", response_model=QueryResponse)
async def query(request: QueryRequest):
    """Answer a question using the RAG pipeline."""
    trace_id = str(uuid.uuid4())
    log = logger.bind(trace_id=trace_id, question=request.question)
    log.info("processing_query")

    answer, citations = await answer_question(request.question, top_k=request.top_k)
    log.info("query_answered", citation_count=len(citations))

    return QueryResponse(
        answer=answer,
        citations=[],
        trace_id=trace_id,
    )
